python/ray/data/datasource/file_based_datasource.py

```diff
@@ -285,6 +285,22 @@ def write(
         **write_args,
     ) -> WriteResult:
         """Write blocks for a file-based datasource."""
+        file_format = self._FILE_EXTENSION
+        if isinstance(file_format, list):
+            file_format = file_format[0]
+
+        builder = DelegatingBlockBuilder()
+        for block in blocks:
+            builder.add_block(block)
+        block = builder.build()
```

Comment:
Not a new issue from this PR. But I'm wondering why we need to merge the input blocks into a big one here. why not just write the input blocks one by one?

Reply:
i'm not sure, but looking at the code and similar usages, my guess is that it has something to do with getting the correct the BlockMetadata of the resulting output.

Reply:
(this comment doesn't block the PR)
Where is code that generates the block metadata of the output? It seems that the block variable is only used locally.

Reply:
i think it may be being used here: https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L2789, but the `datasource.on_write_complete` method seems to just be a pass through...

Reply:
ok. I think this is something we can optimize later. I remember someone said that current the write op will output all the input data. I think we can change it to only output some metadata (e.g, written rows/bytes).